### PR TITLE
All the LINKS are belong to us

### PR DESCRIPTION
**Pull request by np**
This PR depend on #120 so for now it compares to 2016-02-25-np-refactor-ideas-topics.

### CODE REVIEW

File: src/Data/UriPath.hs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+import Thentos.Prelude
+import Data.String.Conversions
+
+import qualified Data.Text as ST
+import qualified Lucid
+
+newtype UriPart = SlashFreeUriPart { fromUriPart :: ST }
+
+instance IsString UriPart where
```
Comment (andorp) [inline src/Data/UriPath.hs]: Why do we need `IsString`?

File: src/Data/UriPath.hs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    fromString s
+        | '/' `elem` s = error $ "UriPart.fromString: unexpected literal '/' in URI part: " <> s
```
Comment (andorp) [inline src/Data/UriPath.hs]: I have concerns about the `error`

**Comment by andorp** [comment]
Two minor comments. Otherwise LGTM

**Comment by fisx** [comment]
`IsString` is for `UriPath` literals; the error is to force use of `(</>)` for path segment separation.

**Comment by fisx** [comment]
@np, did you take a look at `uri-bytestring` (#31)?  the first commit looks like you are re-implementing this?

File: src/Data/UriPath.hs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    mempty = DiffUriParts id
+
+    DiffUriParts ps `mappend` DiffUriParts qs = DiffUriParts (ps . qs)
+
+infixl 7 </>
+
+(</>) :: UriPath -> UriPart -> UriPath
+DiffUriParts ps </> p = DiffUriParts (ps . (p :))
+
+instance IsString UriPath where
```
Comment (fisx) [inline src/Data/UriPath.hs]: This makes it legal to write:

```haskell
p = "no/slashes/" </> "ohno"
```

I don't want to change this, just an observation.

**Comment by fisx** [comment]
More thoughts, good and bad mixed:

- I added a few trivial changes.
- Now that I've seen the entire branch, I have to admit it looks very differnt from `uri-bytestring`.
- I noticed that `UriPath` concatenation allows anybody now to construct invalid paths from valid ones.
- Overall I'm worried about over-engineering.  It's not easy to see your way through this api.
- Tests would be great (especially in light of the above).

In conclusion, LGTM!  I enjoyed the mind-boggling, and I'm looking forward to seeing how this will help us and/or bite us in the future.  :-)

Merging.  Thanks!

File: src/Data/UriPath.hs (hunk 2 above)
Comment (np) [inline src/Data/UriPath.hs]: Better would be to statically check the string literals.

Another option would be to use URI encoding.

Comment (andorp) [inline src/Data/UriPath.hs]: For the time being, we can have it this way. If we encounter issues coming from here, we will change it.

**Comment by np** [comment]
@fisx I didn't look thoroughly into uri-bytestring. The module was short and easy to setup, my goal was to avoid using an alias to ST.

What I don't like:

* UriPath.fromString: this was to support the long static paths for TopStatic.
* Absolute vs relative

Improvements:

* UriPath.fromString: we could at least do some cleaning of leading and trailing slashes.
* Absolute vs relative: it might be worth trying to index UriPath by a type-level Boolean.
  The Monoid instance of UriPath could be restricted to relative.
* The best solution should make use of the routing table.
* I'm not sure the `Space` layer is really helping, for instance the page types are all considered top-level.
  For instance `U.ViewIdea space ideaId` instead of `U.Space space (U.ViewIdea ideaId)`.

Relative paths are supposed to help with relocation, but imagine that for some reason we want to invert the nesting of some paths/captures then our data types reflecting the old scheme would look inconsistent.

My conclusion is that safe links really have a benefit for us and that relative paths add unneeded complexity.
